server/src/main/java/io/druid/client/selector/ServerSelector.java

```diff
@@ -62,59 +67,96 @@ public void addServerAndUpdateSegment(
   {
     synchronized (this) {
       this.segment.set(segment);
-      int priority = server.getServer().getPriority();
-      Set<QueryableDruidServer> priorityServers = servers.computeIfAbsent(priority, p -> new HashSet<>());
+      Set<QueryableDruidServer> priorityServers;
+      if (server.getServer().getType() == ServerType.REALTIME) {
+        priorityServers = realtimeServers.computeIfAbsent(
+            server.getServer().getPriority(), p -> new HashSet<>());
+      } else {
+        priorityServers = historicalServers.computeIfAbsent(
+            server.getServer().getPriority(), p -> new HashSet<>());
+      }
       priorityServers.add(server);
     }
   }
 
   public boolean removeServer(QueryableDruidServer server)
   {
     synchronized (this) {
-      int priority = server.getServer().getPriority();
-      Set<QueryableDruidServer> priorityServers = servers.get(priority);
+      Set<QueryableDruidServer> priorityServers;
+      if (server.getServer().getType() == ServerType.REALTIME) {
+        priorityServers = realtimeServers.computeIfAbsent(
```

Comment:
It should be `computeIfPreset()`, similar to https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/server/coordinator/cost/ClusterCostCache.java#L71-L74